File: src/web-ui/src/infrastructure/api/service-api/UserSecretsAPI.ts
/**
 * UserSecretsAPI — settings CRUD for {{name}} secret variables.
 * List responses never include values.
 */

import { api } from '../index';
import { createTauriCommandError } from '../errors/TauriCommandError';

export interface UserSecretSummary {
  name: string;
  updatedAt: number;
}

export class UserSecretsAPI {
  async list(): Promise<UserSecretSummary[]> {
    try {
      return await api.invoke<UserSecretSummary[]>('list_user_secrets');
    } catch (error) {
      throw createTauriCommandError('list_user_secrets', error);
    }
  }

  async upsert(name: string, value: string): Promise<UserSecretSummary> {
    try {
      return await api.invoke<UserSecretSummary>('upsert_user_secret', {
        request: { name, value },
      });
    } catch (error) {
      throw createTauriCommandError('upsert_user_secret', error, { name });
    }
  }

  async delete(name: string): Promise<boolean> {
    try {
      return await api.invoke<boolean>('delete_user_secret', {
        request: { name },
      });
    } catch (error) {
      throw createTauriCommandError('delete_user_secret', error, { name });
    }
  }
}

export const userSecretsAPI = new UserSecretsAPI();

File: src/apps/desktop/src/api/user_secrets_api.rs
//! User secret variables — settings CRUD (values never returned).

use bitfun_core::user_secrets::{
    delete_user_secret as delete_user_secret_inner, list_user_secrets as list_user_secrets_inner,
    upsert_user_secret as upsert_user_secret_inner,
};
use bitfun_core_types::{UserSecretSummary, UserSecretUpsert};
use serde::Deserialize;

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpsertUserSecretRequest {
    pub name: String,
    pub value: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DeleteUserSecretRequest {
    pub name: String,
}

#[tauri::command]
pub async fn list_user_secrets() -> Result<Vec<UserSecretSummary>, String> {
    list_user_secrets_inner().await.map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn upsert_user_secret(
    request: UpsertUserSecretRequest,
) -> Result<UserSecretSummary, String> {
    upsert_user_secret_inner(UserSecretUpsert {
        name: request.name,
        value: request.value,
    })
    .await
    .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn delete_user_secret(request: DeleteUserSecretRequest) -> Result<bool, String> {
    delete_user_secret_inner(&request.name)
        .await
        .map_err(|e| e.to_string())
}
